File: Source/ARTMessageOperation.m
#import "ARTMessageOperation.h"
#import "ARTMessageOperation+Private.h"

@implementation ARTMessageOperation

- (instancetype)initWithClientId:(NSString *)clientId descriptionText:(NSString *)descriptionText metadata:(NSDictionary<NSString *, NSString *> *)metadata {
    self = [super init];
    if (self) {
        _clientId = clientId;
        _descriptionText = descriptionText;
        _metadata = metadata;
    }
    return self;
}

- (id)copyWithZone:(NSZone *)zone {
    ARTMessageOperation *operation = [[[self class] allocWithZone:zone] init];
    operation->_clientId = self.clientId;
    operation->_descriptionText = self.descriptionText;
    operation->_metadata = self.metadata;
    return operation;
}

- (NSString *)description {
    NSMutableString *description = [NSMutableString stringWithFormat:@"<%@: %p> {\n", self.class, self];
    [description appendFormat:@" clientId: %@,\n", self.clientId];
    [description appendFormat:@" descriptionText: %@,\n", self.descriptionText];
    [description appendFormat:@" metadata: %@\n", self.metadata];
    [description appendFormat:@"}"];
    return description;
}

- (void)writeToDictionary:(NSMutableDictionary<NSString *, id> *)dictionary {
    if (self.clientId) {
        dictionary[@"clientId"] = self.clientId;
    }
    if (self.descriptionText) {
        dictionary[@"description"] = self.descriptionText;
    }
    if (self.metadata) {
        dictionary[@"metadata"] = self.metadata;
    }
}

+ (nullable instancetype)createFromDictionary:(nonnull NSDictionary *)dictionary {
    NSAssert(false, @"Not implemented");
    return nil;
}

- (nullable instancetype)initWithDictionary:(nonnull NSDictionary *)dictionary {
    NSAssert(false, @"Not implemented");
    return nil;
}

@end

File: Sources/AblySwift/ARTMessageOperation.swift
import Foundation

// swift-migration: original location ARTMessageOperation.h, line 8 and ARTMessageOperation.m, line 4
/// An interface outlining the optional `ARTMessageOperation` object which resides in an `ARTMessage` object. This is populated within the `ARTMessage` object when the message is an update or delete operation.
public class ARTMessageOperation: NSObject {
    
    // swift-migration: original location ARTMessageOperation.h, line 10
    public var clientId: String?
    
    // swift-migration: original location ARTMessageOperation.h, line 11
    public var descriptionText: String?
    
    // swift-migration: original location ARTMessageOperation.h, line 12
    public var metadata: [String: String]?
    
    public override init() {
        super.init()
    }
    
    // MARK: - Private Methods
    
    // swift-migration: original location ARTMessageOperation+Private.h, line 9 and ARTMessageOperation.m, line 6
    /// Serialize the Operation object
    internal func writeToDictionary(_ dictionary: inout [String: Any]) {
        if let clientId = self.clientId {
            dictionary["clientId"] = clientId
        }
        if let descriptionText = self.descriptionText {
            dictionary["description"] = descriptionText
        }
        if let metadata = self.metadata {
            dictionary["metadata"] = metadata
        }
    }
    
    // swift-migration: original location ARTMessageOperation+Private.h, line 12 and ARTMessageOperation.m, line 18
    /// Deserialize an Operation object from a NSDictionary object
    internal static func createFromDictionary(_ jsonObject: [String: Any]) -> ARTMessageOperation {
        let operation = ARTMessageOperation()
        if let clientId = jsonObject["clientId"] as? String {
            operation.clientId = clientId
        }
        if let description = jsonObject["description"] as? String {
            operation.descriptionText = description
        }
        
        if let metadata = jsonObject["metadata"] as? [String: String] {
            operation.metadata = metadata
        }
        
        return operation
    }
}